src/Interpreters/Streaming/Aggregator.cpp

```diff
@@ -1480,34 +1482,53 @@ Block Aggregator::insertResultsIntoColumns(PaddedPODArray<AggregateDataPtr> & pl
     return finalizeBlock(params, getHeader(/*final=*/ true), std::move(out_cols), /*final=*/ true, places.size());
 }
 
-template <typename Method, typename Table>
-Block NO_INLINE Aggregator::convertToBlockImplFinal(
-    Method & method, Table & data, Arena * arena, Arenas & aggregates_pools, size_t rows, bool clear_states, ConvertType type) const
+template <bool return_single_block, typename Method, typename Table>
+Aggregator::ConvertToBlockRes<return_single_block> NO_INLINE
+Aggregator::convertToBlockImplFinal(
+    Method & method, Table & data, Arena * arena, Arenas & aggregates_pools, size_t, bool clear_states, ConvertType type) const
 {
+    /// +1 for nullKeyData, if `data` doesn't have it - not a problem, just some memory for one excessive row will be preallocated
+    const size_t max_block_size = (return_single_block ? data.size() : std::min(params.max_block_size, data.size())) + 1;
     constexpr bool final = true;
-    auto out_cols = prepareOutputBlockColumns(params, aggregate_functions, getHeader(final), aggregates_pools, final, rows);
+    ConvertToBlockRes<return_single_block> res;
 
-    if constexpr (Method::low_cardinality_optimization)
+    std::optional<OutputBlockColumns> out_cols;
```

Comment:
Use a plain `OutputBlockColumns` is good which we can move away its underlying data with out a explicit reset ?